components/net/clientstorage/thread.rs

```diff
@@ -0,0 +1,168 @@
+                let _ = msg_sender.send(result.unwrap());
+            }),
+        );
+
+        loop {
+            let msg = self.msg_receiver.recv().unwrap();
+
+            self.recv_thread_message(msg);
+
+            if self.exiting.get() {
+                break;
+            }
+        }
+    }
+
+    fn recv_thread_message(self: &Rc<Self>, msg: ClientStorageThreadMsg) {
+        match msg {
+            ClientStorageThreadMsg::TestConstructor {
+                child_to_parent_receiver,
+                parent_to_child_sender,
+                sender,
+            } => {
+                let actor_id =
+                    self.recv_test_constructor(child_to_parent_receiver, parent_to_child_sender);
+                let _ = sender.send(actor_id);
+            },
+
+            ClientStorageThreadMsg::Routed(msg) => {
+                self.recv_routed_message(msg);
+            },
+
+            ClientStorageThreadMsg::Exit(sender) => {
+                self.recv_exit();
+                let _ = sender.send(());
+            },
+        }
+    }
+
+    fn recv_test_constructor(
+        self: &Rc<Self>,
+        child_to_parent_receiver: IpcReceiver<ClientStorageRoutedMsg>,
+        parent_to_child_sender: IpcSender<ClientStorageRoutedMsg>,
+    ) -> ClientStorageActorId {
+        let msg_sender = self.msg_sender.clone();
+
+        let global_id = self.next_global_id.get();
+        self.next_global_id.set(global_id + 1);
+
+        let actor_id = ClientStorageActorId {
+            global_id,
+            local_id: 0,
+        };
+
+        // TODO: Switch to `add_typed_route_with_close_callback` once available,
+        // and clean up associated actors when the channel closes unexpectedly.
+        ROUTER.add_typed_route(
+            child_to_parent_receiver,
+            Box::new(move |result| {
+                let msg = result.unwrap();
+                let _ = msg_sender.send(ClientStorageThreadMsg::Routed(msg));
+            }),
+        );
+
+        let actor = ClientStorageTestParent::new();
+
+        actor.bind(Rc::clone(self), parent_to_child_sender, actor_id);
+
+        actor_id
+    }
+
+    fn recv_routed_message(self: &Rc<Self>, msg: ClientStorageRoutedMsg) {
+        let actor = self.actors.borrow().get(&msg.actor_id).unwrap().clone();
+
+        match (actor, msg.data) {
+            (
+                ClientStorageParent::ClientStorageTest(actor),
+                ClientStorageMixedMsg::ClientStorageTest(msg),
+            ) => {
+                actor.recv_message(msg);
+            },
+
+            (
+                ClientStorageParent::ClientStorageTestCursor(actor),
+                ClientStorageMixedMsg::ClientStorageTestCursor(msg),
+            ) => {
+                actor.recv_message(msg);
+            },
+
+            _ => {},
+        }
+    }
+
+    fn recv_exit(self: &Rc<Self>) {
+        self.exiting.set(true);
+    }
+
+    pub fn register_actor(
+        self: &Rc<Self>,
```

Comment:
Speaking of, why use a rc instead of a box here

Reply:
Never mind, I see it now, I think this does force all nodes to be in the same thread, which is suboptimal if we are trying to parallelize things.